# Dataset Preparation
- Inputs:
    - American Housing Survey (AHS) for survey years 2015, 2017, 2019, 2021, 2023.
    - HUD Income Limits for Select Metro Areas and Years (2015, 2017, 2019, 2021, 2023)
        - 12060: Atlanta-Sandy Springs-Roswell, GA
        - 14460: Boston-Cambridge-Newton, MA-NH
        - 16980: Chicago-Naperville-Elgin, IL-IN-WI
        - 19100: Dallas-Fort Worth-Arlington, TX
        - 19820: Detroit-Warren-Dearborn, MI
        - 26420: Houston-The Woodlands-Sugar Land, TX
        - 31080: Los Angeles-Long Beach-Anaheim, CA
        - 33100: Miami-Fort Lauderdale-West Palm Beach, FL
        - 35620: New York-Newark-Jersey City, NY-NJ-PA
        - 37980: Philadelphia-Camden-Wilmington, PA-NJ-DE-MD
        - 38060: Phoenix-Mesa-Scottsdale, AZ
        - 40140: Riverside-San Bernardino-Ontario, CA
        - 41860: San Francisco-Oakland-Hayward, CA
        - 42660: Seattle-Tacoma-Bellevue, WA
        - 47900: Washington-Arlington-Alexandria, DC-VA-MD-WV
- Output:
    - A panel dataset (2015 - 2023) that includes an `'AMI'` variable that categorizes each observation/household (identified by the `'CONTROL'` variable) by Area Median Income level based on the corresponding Income Limit thresholds for that Metropolitan and fiscal year defined by the U.S. Department of Housing and Urban Development (HUD). 

This notebook accomplishes the following parts of the workflow/analysis:
1. Complies a longitudinal dataset.
2. Isolates the AHS data to renter and owner-occupied housing units.
3. Identfies unique renter/owner occupied housing units by AMI level for the 15 largest metropolitan areas identified in the AHS:
    - Above LI: >80% AMI
    - Low-Income (LI): <= 80% AMI
    - Very Low-Income (VLI): <= 50% AMI
    - Extemely Low-Income (ELI): <= 30% AMI

In [1]:
#Importing Libraries
import requests
from io import StringIO

import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
from matplotlib.ticker import PercentFormatter
import seaborn as sns

## Data Access
The AHS Public Use Files (PUFs) can be downloaded from census.gov:

- [2015 AHS National PUF](https://www.census.gov/programs-surveys/ahs/data/2015/ahs-2015-public-use-file--puf-/ahs-2015-national-public-use-file--puf-.html)
- [2017 AHS National PUF](https://www.census.gov/programs-surveys/ahs/data/2017/ahs-2017-public-use-file--puf-/ahs-2017-national-public-use-file--puf-.html)
- [2019 AHS National PUF](https://www.census.gov/programs-surveys/ahs/data/2019/ahs-2019-public-use-file--puf-/ahs-2019-national-public-use-file--puf-.html)
- [2021 AHS National PUF](https://www.census.gov/programs-surveys/ahs/data/2021/ahs-2021-public-use-file--puf-/ahs-2021-national-public-use-file--puf-.html)
- [2023 AHS National PUF](https://www.census.gov/programs-surveys/ahs/data/2023/ahs-2023-public-use-file--puf-/ahs-2023-national-public-use-file--puf-.html)

Note: This analysis uses the PUF "Flat File."

In [2]:
#Selected relevant variables for the analysis
columns_15 = ['CONTROL', 'WEIGHT', 'OMB13CBSA', 'VACANCY', 'VACMONTHS', 'TENURE', 'NUMPEOPLE', 'HINCP', 'RENT', 'TOTHCAMT',
           'HUDSUB', 'RENTSUB', 'RENTCNTRL', 'BLD', 'YRBUILT', 'BEDROOMS', 'HHMOVE']

columns = ['CONTROL', 'WEIGHT', 'OMB13CBSA', 'VACANCY', 'VACMONTHS', 'TENURE', 'NUMPEOPLE', 'HINCP', 'RENT', 'TOTHCAMT',
           'HUDSUB', 'RENTSUB', 'RENTCNTRL', 'BLD', 'YRBUILT', 'BEDROOMS', 'SAMEHHLD', 'HHMOVE']

#Loading 2015 AHS METRO PUF w/ selected columns
df_15 = pd.read_csv('data/ahs/ahs2015n.csv',
                    usecols=columns_15)

#Loading 2017 AHS METRO PUF w/ selected columns
df_17 = pd.read_csv('data/ahs/ahs2017n.csv',
                    usecols=columns)

#Loading 2019 AHS METRO PUF w/ selected columns
df_19 = pd.read_csv('data/ahs/ahs2019n.csv',
                    usecols=columns)
#Loading 2021 AHS METRO PUF w/ selected columns
df_21 = pd.read_csv('data/ahs/ahs2021n.csv',
                    usecols=columns)

#Loading 2023 AHS METRO PUF w/ selected columns
df_23 = pd.read_csv('data/ahs/ahs2023n.csv',
                    usecols=columns)

In [3]:
#Storing the DataFrames in a dictionary with full survey years as keys
dfs_by_year = {
    2023: df_23,
    2021: df_21,
    2019: df_19,
    2017: df_17,
    2015: df_15
}

#Previewing each df
for year, df in dfs_by_year.items():
    print(f'--- AHS {year} ---')
    print(df.head(), '\n')

--- AHS 2023 ---
      CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA       WEIGHT   BLD  \
0  '11000002'  1600   '-6'      '-6'     '8'   '99998'   813.890194  '03'   
1  '11000003'   840    '2'      '-6'     '8'   '99998'   581.103231  '03'   
2  '11000005'    -6    '1'      '-6'    '-6'   '99998'  7335.965001  '02'   
3  '11000006'    -6    '1'      '-6'    '-6'   '99998'  6562.865941  '02'   
4  '11000008'   800    '2'      '-6'     '8'   '99998'  1490.800600  '06'   

   HHMOVE  NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY  \
0      -6         -6     '-6'     1980         4          0   '-6'    '02'   
1    2023          3      '4'     1970         2         -6    '3'    '-6'   
2    1995          2      '1'     1970         4         -6   '-6'    '-6'   
3    2019          3      '4'     1970         3         -6   '-6'    '-6'   
4    2019          1      '4'     1920         1         -6    '3'    '-6'   

    HINCP  TOTHCAMT  
0      -6        -6  
1   480

In [4]:
#Checking the shape/info of each DataFrame
for year, df in dfs_by_year.items():
    print(f'\n--- AHS {year} ---')
    print('Shape:', df.shape)
    df.info()


--- AHS 2023 ---
Shape: (55669, 18)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55669 entries, 0 to 55668
Data columns (total 18 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   CONTROL    55669 non-null  object 
 1   RENT       55669 non-null  int64  
 2   TENURE     55669 non-null  object 
 3   RENTCNTRL  55669 non-null  object 
 4   RENTSUB    55669 non-null  object 
 5   OMB13CBSA  55669 non-null  object 
 6   WEIGHT     55669 non-null  float64
 7   BLD        55669 non-null  object 
 8   HHMOVE     55669 non-null  int64  
 9   NUMPEOPLE  55669 non-null  int64  
 10  SAMEHHLD   55669 non-null  object 
 11  YRBUILT    55669 non-null  int64  
 12  BEDROOMS   55669 non-null  int64  
 13  VACMONTHS  55669 non-null  int64  
 14  HUDSUB     55669 non-null  object 
 15  VACANCY    55669 non-null  object 
 16  HINCP      55669 non-null  int64  
 17  TOTHCAMT   55669 non-null  int64  
dtypes: float64(1), int64(8), object(9)
memory usage: 

## Setting up and Preparing DataFrames for Longitudinal Analysis
- Adding `SRVYEAR` variable/column to identify the year of the AHS dataset that will be concatenated later in the analysis
- Using `.concat()` to stack the DataFrame vertically (long format)
- Stripping the extra `''` from the sting values 

In [5]:
#Add the survey year as a new column in each DataFrame
dfs_by_year = {
    year: df.assign(SRVYEAR=int(year)) for year, df in dfs_by_year.items()
}

In [6]:
#Checking the DataFrames to see if the column was added
for year, df in dfs_by_year.items():
    print(f'--- AHS {year} ---')
    print(df.head(), '\n')

--- AHS 2023 ---
      CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA       WEIGHT   BLD  \
0  '11000002'  1600   '-6'      '-6'     '8'   '99998'   813.890194  '03'   
1  '11000003'   840    '2'      '-6'     '8'   '99998'   581.103231  '03'   
2  '11000005'    -6    '1'      '-6'    '-6'   '99998'  7335.965001  '02'   
3  '11000006'    -6    '1'      '-6'    '-6'   '99998'  6562.865941  '02'   
4  '11000008'   800    '2'      '-6'     '8'   '99998'  1490.800600  '06'   

   HHMOVE  NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY  \
0      -6         -6     '-6'     1980         4          0   '-6'    '02'   
1    2023          3      '4'     1970         2         -6    '3'    '-6'   
2    1995          2      '1'     1970         4         -6   '-6'    '-6'   
3    2019          3      '4'     1970         3         -6   '-6'    '-6'   
4    2019          1      '4'     1920         1         -6    '3'    '-6'   

    HINCP  TOTHCAMT  SRVYEAR  
0      -6        -6 

In [7]:
#checking the average of the `WEIGHT` variables. Each year should average ~2,000
avg23 = np.average(df_23['WEIGHT'])
print(avg23)

2610.8720472795485


### Concatenating the 2015, 2017, 2019, 2021, and 2023 AHS PUFs to produce a **long-format panel dataset**.
This will allow:
- **Track units over time** --> This requires **multiple rows per unit** (i.e., one row per unit * per year).
- **Identify the same unit** --> This is already possible with the `CONTROL` variable.
- **Compare status across years** (e.g., 2015 vs. 2023) --> This is easiest when **filtering by year** can be cleanly, which is what the `SRVYEAR` column enables.

In [8]:
#Stack all years into a single long-form DataFrame
ahs_long_df = pd.concat(dfs_by_year.values(), ignore_index=True)

In [9]:
#Inspecting the shape and preview the concatenated DataFrames
print(ahs_long_df.shape)
print(ahs_long_df.info())
print(ahs_long_df.head())
print(ahs_long_df['SRVYEAR'].value_counts())

(319240, 19)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 319240 entries, 0 to 319239
Data columns (total 19 columns):
 #   Column     Non-Null Count   Dtype  
---  ------     --------------   -----  
 0   CONTROL    319240 non-null  object 
 1   RENT       319240 non-null  int64  
 2   TENURE     319240 non-null  object 
 3   RENTCNTRL  319240 non-null  object 
 4   RENTSUB    319240 non-null  object 
 5   OMB13CBSA  319240 non-null  object 
 6   WEIGHT     319240 non-null  float64
 7   BLD        319240 non-null  object 
 8   HHMOVE     319240 non-null  int64  
 9   NUMPEOPLE  319240 non-null  int64  
 10  SAMEHHLD   249747 non-null  object 
 11  YRBUILT    319240 non-null  int64  
 12  BEDROOMS   319240 non-null  int64  
 13  VACMONTHS  319240 non-null  int64  
 14  HUDSUB     319240 non-null  object 
 15  VACANCY    319240 non-null  object 
 16  HINCP      319240 non-null  int64  
 17  TOTHCAMT   319240 non-null  int64  
 18  SRVYEAR    319240 non-null  int64  
dtypes: float64

In [10]:
#Checking for unique unit-year combinations, the lines of code should produce the same output
print(len(ahs_long_df))
print(ahs_long_df[['CONTROL', 'SRVYEAR']].drop_duplicates().shape[0])

319240
319240


In [11]:
#Ensuring no columns were dropped
print(ahs_long_df.columns)

Index(['CONTROL', 'RENT', 'TENURE', 'RENTCNTRL', 'RENTSUB', 'OMB13CBSA',
       'WEIGHT', 'BLD', 'HHMOVE', 'NUMPEOPLE', 'SAMEHHLD', 'YRBUILT',
       'BEDROOMS', 'VACMONTHS', 'HUDSUB', 'VACANCY', 'HINCP', 'TOTHCAMT',
       'SRVYEAR'],
      dtype='object')


In [12]:
#Looking for missing or duplicated `CONTROL`-year rows
dupes = ahs_long_df.duplicated(subset=["CONTROL", "SRVYEAR"], keep=False)
print("Duplicated CONTROL + SRVYEAR rows:", dupes.sum())

Duplicated CONTROL + SRVYEAR rows: 0


In [13]:
#Checking counts by year to confirm stacking
ahs_long_df["SRVYEAR"].value_counts().sort_index()

SRVYEAR
2015    69493
2017    66752
2019    63185
2021    64141
2023    55669
Name: count, dtype: int64

##### Stripping the extra quotation marks `''` from the sting values of the entire dataset to make coding a cleaner/easier

In [14]:
ahs_long_df[ahs_long_df.select_dtypes(include='object')
            .columns] = ahs_long_df.select_dtypes(include='object').apply(lambda col: col.str.strip("'"))
ahs_long_df.head()

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA       WEIGHT BLD  HHMOVE  \
0  11000002  1600     -6        -6       8     99998   813.890194  03      -6   
1  11000003   840      2        -6       8     99998   581.103231  03    2023   
2  11000005    -6      1        -6      -6     99998  7335.965001  02    1995   
3  11000006    -6      1        -6      -6     99998  6562.865941  02    2019   
4  11000008   800      2        -6       8     99998  1490.800600  06    2019   

   NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY   HINCP  \
0         -6       -6     1980         4          0     -6      02      -6   
1          3        4     1970         2         -6      3      -6   48000   
2          2        1     1970         4         -6     -6      -6  292500   
3          3        4     1970         3         -6     -6      -6   56000   
4          1        4     1920         1         -6      3      -6   36000   

   TOTHCAMT  SRVYEAR  
0        -6     2023  
1      1093     2023  
2       810     2023  
3       489     2023  
4       845     2023

### Creating new DataFrame to exclude "Units occupied without payment of rent"
- The new `ahs_panel_df` DataFrame will include owner-occupied units (`'1'`), renter--occupied units (`'2'`), and not applicable/vacant units (`'-6'`).

In [15]:
#Examining `TENURE` variable
ahs_long_df['TENURE'].value_counts()

TENURE
1     162666
2     111084
-6     41729
3       3761
Name: count, dtype: int64

The new DataFrame will consist of:
- Owner-occupied housing units (`'1'`): 162,666
- Renter-occupied housing units (`'2'`): 111,084
- Not Applicable (`'-6'`): 41,729 -- _need this for vacant unit analysis_

The new DataFrame will disregard:
- Units occupied w/o payment of rent (`'3'`): 3,761

In [16]:
#Looking at the characteristic of 'HINCP' to see if "N" values exists in the DataFrame or if its just '-6' values
print(ahs_long_df['HINCP'].dtype)
print(ahs_long_df['HINCP'].unique())

#Number of unique values in the df
print(ahs_long_df['HINCP'].nunique())

int64
[    -6  48000 292500 ...    364 327100 111660]
16891


In [17]:
#Looking at the number of renter & owner occupied housing units that have "not applicable" ('-6') values for household income ('HINCP')
ahs_long_df[(ahs_long_df['TENURE'] == '1') | (ahs_long_df['TENURE'] == '2')].groupby('HINCP').size().get(-6,0)

0

In [18]:
ahs_long_df[((ahs_long_df['TENURE'] == '1') | (ahs_long_df['TENURE'] == '2')) & (ahs_long_df['HINCP'] == -6)].shape[0]

0

In [19]:
#Checking any NA/-6 values among renter houhseholds in the df
grouped = ahs_long_df[(ahs_long_df['TENURE'] == '1') | (ahs_long_df['TENURE'] == '2')].groupby('HINCP').size()

#Check if -6 is in the index before using .loc
if -6 in grouped.index:
    count_minus_6 = grouped.loc[-6]
else:
    count_minus_6 = 0

print("Number of owner/renter-occupied housing units with HINCP == -6:", count_minus_6)

Number of owner/renter-occupied housing units with HINCP == -6: 0


All owner/renter observations have a dollar amount within the `'HINCP'` variable.

In [20]:
#Looking at sum of renter/owner-occupied/vacant housing units
rent_own_count = ((ahs_long_df['TENURE'] == '1') | (ahs_long_df['TENURE'] == '2') | (ahs_long_df['TENURE'] == '-6')).sum()
print(rent_own_count)

315479


In [21]:
#New df consisting of owner/renter-occupied + vacant housing units
ahs_panel_df = ahs_long_df[ahs_long_df['TENURE'] != '3']

#Checking if `ahs_panel_df` was filtered correctly
len(ahs_panel_df)

315479

Owner-occupied housing units ('1'): 162,666 + Renter-occupied housing units ('2'): 111,084 + Not Applicable/Vacant ('6'): 41,729 = 315,479

### Examinning the household income `'HINCP'` and number of people in the unit `'NUMPEOPLE'` variables for renter/owner-occupied housing units
- These variables are critical for matching observations in the AHS to the correct Area Median Income (AMI) category based on HUD's Income Limit thresholds.

In [22]:
pd.options.display.float_format = '{:.2f}'.format #changing numeric format to see the dollar amount more clearly

ahs_panel_df.groupby('TENURE')[['HINCP', 'NUMPEOPLE']].describe()

HINCP                                                            \
           count      mean       std       min      25%      50%       75%   
TENURE                                                                       
-6      41729.00     -6.00      0.00     -6.00    -6.00    -6.00     -6.00   
1      162666.00 110403.59 138828.23  -5000.00 40000.00 78000.00 135000.00   
2      111084.00  52883.59  77385.25 -10000.00 14100.00 34200.00  68000.00   

                  NUMPEOPLE                                           
              max     count  mean  std   min   25%   50%   75%   max  
TENURE                                                                
-6          -6.00  41729.00 -6.00 0.00 -6.00 -6.00 -6.00 -6.00 -6.00  
1      5786000.00 162666.00  2.58 1.43  1.00  2.00  2.00  3.00 18.00  
2      6445000.00 111084.00  2.27 1.48  1.00  1.00  2.00  3.00 19.00

In [23]:
#Examining the the oultiers
ahs_panel_df.groupby('TENURE')[['HINCP', 'NUMPEOPLE']].quantile([0, .01, .2,.4,.6,.8,.99])

HINCP  NUMPEOPLE
TENURE                          
-6     0.00     -6.00      -6.00
       0.01     -6.00      -6.00
       0.20     -6.00      -6.00
       0.40     -6.00      -6.00
       0.60     -6.00      -6.00
       0.80     -6.00      -6.00
       0.99     -6.00      -6.00
1      0.00  -5000.00       1.00
       0.01      0.00       1.00
       0.20  32300.00       1.00
       0.40  61200.00       2.00
       0.60  97000.00       3.00
       0.80 153000.00       4.00
       0.99 592415.00       7.00
2      0.00 -10000.00       1.00
       0.01      0.00       1.00
       0.20  11400.00       1.00
       0.40  25000.00       1.00
       0.60  45000.00       2.00
       0.80  79800.00       3.00
       0.99 315000.00       7.00

In [24]:
#Looking at the number of renter/owner-occpuied houing units with negative incomes
ahs_panel_df.loc[ahs_panel_df["HINCP"] < 0].groupby('TENURE')[['HINCP', 'NUMPEOPLE']].describe()

HINCP                                                                \
          count     mean     std       min      25%      50%      75%     max   
TENURE                                                                          
-6     41729.00    -6.00    0.00     -6.00    -6.00    -6.00    -6.00   -6.00   
1         10.00 -2592.40 1887.42  -5000.00 -4440.00 -2450.00 -1175.00   -4.00   
2          5.00 -5160.00 3260.06 -10000.00 -5000.00 -5000.00 -5000.00 -800.00   

       NUMPEOPLE                                           
           count  mean  std   min   25%   50%   75%   max  
TENURE                                                     
-6      41729.00 -6.00 0.00 -6.00 -6.00 -6.00 -6.00 -6.00  
1          10.00  2.90 3.63  1.00  1.00  2.00  2.75 13.00  
2           5.00  1.80 1.10  1.00  1.00  1.00  3.00  3.00

In [25]:
ahs_panel_df.loc[ahs_panel_df["HINCP"] < 0, 'TENURE'].value_counts()

TENURE
-6    41729
1        10
2         5
Name: count, dtype: int64

There are 15 observations in the sample that have negative values for `'HINCP'`. These observations will not be removed from the DataFrame since it is possible (or at least, for the American Community Survey it is, that these observations are meaningful if a renter/owner household moves into a unit with no income. Plus, the AHS tracks _**housing units**_(`'CONTROL'`), not households.   

In addition, housing unit occupants are catagorized by Area Median Income (AMI) later in this data preparation notebook, these observations would be catagorized as Extremely Low-Income with no major effect to the data and the panel dataset will have less "gaps" in observations.

## Loading in Income Limits to catagorize housing units by Area Median Income (AMI) defined by the U.S. Department of Housing and Urban Development (HUD) Income Limits

Accessing 2017, 2019, 2021, and 2023 Income Limit data via the API Function from HUD USER (https://www.huduser.gov/portal/dataset/fmr-api.h) and will merge the newly created `staked_il_df` on the CBSA code (`'OMB13CBSA'`) variable to the AHS data and then create a new `'AMI'` categorical variable with the following values:
- Above LI: Above Low-Income
    - Households above the low-income thresholds of 80% Area Median Income (AMI).
    - Households above 80% AMI are not elibigle for federal rental subsidies in most cases.
- LI: Low-Income
    - Renter households with incomes at or below 80% AMI.
- VLI: Very Low-Income
    - Renter Households with incomes at or below 50% AMI.
- ELI: Extremely Low-Income
    - Renter Households with incomes at or below 30% AMI

**NOTE:** HUD _**does not**_ support API requests for years prior to 2017. I will need to download the FY2015 from HUDUSER, isolate the corresponding CBSA/Metro Area, and merge that into the API-requested DataFrame.

A stacked DataFrame will be created with a new column `'il_fiscal_year'` to merge the stacked income limits DataFrame with the longitudinal AHS data to match an household's `'HINCP'` (household income) with the corresponding `'AMI'` category based on the income limits of that fiscal year,

The AHS National PUF only surveys the top 15 largest metropolitan areas. Due to this limitation, the following code will load in HUD's Income Limits data for the following metropolitan areas. AHS observations that are not located within the top 15 largest metropolitan areas (`'99998'` - All other metroplitan areas not in Top 15, and `'99999'` - Not in a metropolitan area) will not have corresponding income limit data once the two datasets are merged amd will be omitted from AMI classification.
- Atlanta-Sandy Springs-Roswell, GA (12060)
- Chicago-Naperville-Elgin, IL-IN-WI (16980)
- Dallas-Fort Worth-Arlington, TX (19100)
- Detroit-Warren-Dearborn, MI (19820)
- Houston-The Woodlands-Sugar Land, TX (26420)
- Philadelphia-Camden-Wilmington, PA-NJ-DE-MD (37980)
- Phoenix-Mesa-Scottsdale, AZ (38060)
- Riverside-San Bernardino-Ontario, CA (40140)
- Washington-Arlington-Alexandria, DC-VA-MD (47900)

**Note:** HUD uses a different geographical definition (`'hud_area_code'`) to establish Income Limits for the metropolitan areas below. HUD Income Limts use "HUD Metro FMR Area" which are different than the AHS geographical definition (`'OMB13CBSA'`) for that area. For the purposes of this analysis, the "central" county of the HUD Metro FMR Area within the AHS defined metropolitan area will be used for the purposes of assigning an AMI catagory for an observation. For example, the AHS-defined "Miami-Fort Lauderdale-West Palm Beach, FL" metropolitan area will use the "Miami-Miami Beach-Kendall, FL HUD Metro FMR Area" since Miami-Dade county uses that FMR area; even though Fort Lauderdale and West Palm Beach are part of the AHS-defined metropolitan area but have different income thresholds for the purpose of establishing HUD Income Limts.
- Boston-Cambridge-Quincy, MA-NH (14460) - Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area (Suffolk County)
- Miami-Fort Lauderdale-West Palm Beach, FL (33100) - Miami-Miami Beach-Kendall, FL HUD Metro FMR Area (Miami-Dade County)
- Los Angeles-Long Beach-Anaheim, CA (31080) - Los Angeles-Long Beach-Glendale, CA HUD Metro FMR Area (Los Angeles County)
- New York-Newark-Jersey City, NY-NJ-PA (35620) - New York, NY HUD Metro FMR Area (New York County)
- San Francisco-Oakland-Hayward, CA (41860) - San Francisco, CA HUD Metro FMR Area (San Francisco County)
- Seattle-Tacoma-Bellevue, WA (42660) - Seattle-Bellevue, WA HUD Metro FMR Area (King County)

In [26]:
#API key
API_KEY = 'API_KEY'

#Identifying metro area
cbsa_codes = [
    'METRO12060M12060', #ATL
    'METRO14460MM1120', #BOS
    'METRO16980M16980', #CHI
    'METRO19100M19100', #DAL
    'METRO19820M19820', #DET
    'METRO26420M26420', #HOU
    'METRO31080MM4480', #LA
    'METRO33100MM5000', #MIA
    'METRO35620MM5600', #NYC
    'METRO37980M37980', #PHI
    'METRO38060M38060', #PHX
    'METRO40140M40140', #RIVERSIDE, CA
    'METRO41860MM7360', #SF
    'METRO42660MM7600', #SEA
    'METRO47900M47900'  #DC
]

#Identifying years for the staked DataFrame
years = [2017, 2019, 2021, 2023] 

#Initialize both dictionaries & storage
metro_dict = {}  
year_dict = {}   
all_rows = []

#API request & loop through CBSA and year
for cbsa_code in cbsa_codes:
    for year in years:
        url = f'https://www.huduser.gov/hudapi/public/il/data/{cbsa_code}?year={year}'
        headers = {'Authorization': f'Bearer {API_KEY}'}
        response = requests.get(url, headers=headers)
    
        json_data = response.json()
        data = json_data.get('data',{})
        
        #Structuring the json into a panda dataframe that is reflective of the excel file avaialbe on the web
        #Parse row
        row = {
            'hud_area_code': cbsa_code,
            'hud_area_name': data.get('area_name'),
            'il_fiscal_year': data.get('year'),
            'median_income': data.get('median_income')
        }
        
        #Add l50 (Very Low), ELI (Exrememly Low), l80 (Low)
        for i in range(1, 9):
            row[f'l50_{i}'] = data.get('very_low', {}).get(f'il50_p{i}')
            row[f'ELI_{i}'] = data.get('extremely_low', {}).get(f'il30_p{i}')
            row[f'l80_{i}'] = data.get('low', {}).get(f'il80_p{i}')
            
        #Add dictionaries
        metro_dict.setdefault(cbsa_code, []).append(row)
        year_dict.setdefault(int(data.get('year')), []).append(row)
        
        #Adding to masterlist
        all_rows.append(row)
        
#Creating the final Dataframe for income limits from 2017-2023 for the target metro areas
metro_staked_il_df = pd.DataFrame(all_rows)
metro_staked_il_df

hud_area_code                                      hud_area_name  \
0   METRO12060M12060  Atlanta-Sandy Springs-Roswell, GA HUD Metro FM...   
1   METRO12060M12060  Atlanta-Sandy Springs-Roswell, GA HUD Metro FM...   
2   METRO12060M12060  Atlanta-Sandy Springs-Roswell, GA HUD Metro FM...   
3   METRO12060M12060  Atlanta-Sandy Springs-Roswell, GA HUD Metro FM...   
4   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
5   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
6   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
7   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
8   METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
9   METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
10  METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
11  METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
12  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
13  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
14  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
15  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
16  METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
17  METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
18  METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
19  METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
20  METRO26420M26420  Houston-The Woodlands-Sugar Land, TX HUD Metro...   
21  METRO26420M26420  Houston-The Woodlands-Sugar Land, TX HUD Metro...   
22  METRO26420M26420  Houston-The Woodlands-Sugar Land, TX HUD Metro...   
23  METRO26420M26420  Houston-The Woodlands-Sugar Land, TX HUD Metro...   
24  METRO31080MM4480  Los Angeles-Long Beach-Glendale, CA HUD Metro ...   
25  METRO31080MM4480  Los Angeles-Long Beach-Glendale, CA HUD Metro ...   
26  METRO31080MM4480  Los Angeles-Long Beach-Glendale, CA HUD Metro ...   
27  METRO31080MM4480  Los Angeles-Long Beach-Glendale, CA HUD Metro ...   
28  METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
29  METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
30  METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
31  METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
32  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
33  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
34  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
35  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
36  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
37  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
38  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
39  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
40  METRO38060M38060                    Phoenix-Mesa-Scottsdale, AZ MSA   
41  METRO38060M38060                    Phoenix-Mesa-Scottsdale, AZ MSA   
42  METRO38060M38060                    Phoenix-Mesa-Scottsdale, AZ MSA   
43  METRO38060M38060                    Phoenix-Mesa-Scottsdale, AZ MSA   
44  METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
45  METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
46  METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
47  METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
48  METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
49  METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
50  METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
51  METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
52  METRO42660MM7600            

In [27]:
#Loading in HUD's 2015 Income Limit Data for target CBSAs

#File path
file_path = 'data/ahs/Section8_Rev.xlsx'

#Using excel rows to identify target CBSAs (1-based indexing)
target_rows_excel = [
    609, #ATL
    2182, #BOS 
    773, #CHI
    3880, #DAL
    2327, #DET
    3924, #HOU
    206, #LA
    525, #MIA
    3123, #NYC
    3561, #PHI
    105, #PHX
    220, #RIVERSIDE, CA
    225, #SF
    4513, #SEA
    482 #DC
]

#Convert to 0-based pandas row indices
target_indices = [i - 1 for i in target_rows_excel]

#Read the file, only desired rows (skip all others)
il_metro_15_df = pd.read_excel(
    file_path,
    header=None,
    skiprows=lambda x: x not in target_indices
)

#Read only the header row separately (row 0) for column names
headers = pd.read_excel(file_path, nrows=0).columns.tolist()

#Assign headers back into DataFrame
il_metro_15_df.columns = headers
il_metro_15_df

State_Alpha    fips2000  State  County           County_Name  \
0           AZ   401399999      4      13       Maricopa County   
1           CA   603799999      6      37    Los Angeles County   
2           CA   606599999      6      65      Riverside County   
3           CA   607599999      6      75  San Francisco County   
4           DC  1100199999     11       1  District of Columbia   
5           FL  1208699999     12      86     Miami-Dade County   
6           GA  1312199999     13     121         Fulton County   
7           IL  1703199999     17      31           Cook County   
8           MA  2502507000     25      25        Suffolk County   
9           MI  2616399999     26     163          Wayne County   
10          NY  3606199999     36      61       New York County   
11          PA  4210199999     42     101   Philadelphia County   
12          TX  4811399999     48     113         Dallas County   
13          TX  4820199999     48     201         Harris County   
14          WA  5303399999     53      33           King County   

             CBSASub                                    Metro_Area_Name  \
0   METRO38060M38060                      Phoenix-Mesa-Glendale, AZ MSA   
1   METRO31100MM4480      Los Angeles-Long Beach, CA HUD Metro FMR Area   
2   METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
3   METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
4   METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
5   METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
6   METRO12060M12060  Atlanta-Sandy Springs-Marietta, GA HUD Metro F...   
7   METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
8   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
9   METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
10  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
11  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
12  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
13  METRO26420M26420  Houston-Baytown-Sugar Land, TX HUD Metro FMR Area   
14  METRO42660MM7600            Seattle-Bellevue, WA HUD Metro FMR Area   

      fips2010  median2015  l50_1  ...  l80_3  l80_4   l80_5   l80_6   l80_7  \
0    401399999       64000  22400  ...  46100  51200   55300   59400   63500   
1    603799999       63000  29050  ...  59800  66400   71750   77050   82350   
2    606599999       60500  21750  ...  44750  49700   53700   57700   61650   
3    607599999      101900  41050  ...  84500  93850  101400  108900  116400   
4   1100199999      109200  38250  ...  61200  68000   73450   78900   84350   
5   1208699999       49900  23700  ...  48750  54150   58500   62850   67150   
6   1312199999       68300  23900  ...  49100  54550   58950   63300   67650   
7   1703199999       76000  26600  ...  54750  60800   65700   70550   75400   
8   2502507000       98500  34500  ...  62750  69700   75300   80900   86450   
9   2616399999       67700  23700  ...  48750  54150   58500   62850   67150   
10  3606199999       63700  30250  ...  62150  69050   74600   80100   85650   
11  4210199999       81100  28400  ...  58450  64900   70100   75300   80500   
12  4811399999       70400  24650  ...  50700  56300   60850   65350   69850   
13  4820199999       69300  24300  ...  49950  55450   59900   64350   68800   
14  5303399999       89600  31400  ...  59250  65800   71100   76350   81600   

     l80_8   MSA      county_town_name            state_name  metro  
0    67600  6200       Maricopa County               Arizona      1  
1    87650  4480    Los Angeles County            California      1  
2    65650  6780      Riverside County            California      1  
3   123900  7360  San Francisco County            California      1  
4    89800  8840  District of Columbia  District of Columbia      1  
5    71500  5000     Miami-D

#### Structuring the FY2015 Income Limit DataFrame to match the API-request DataFrame above.

In [28]:
#Dropping columns that are not relevant 
il_metro_15_df.drop(columns=['State_Alpha', 'fips2000', 'State', 'County', 'County_Name', 'fips2010', 'MSA', 'county_town_name',
                           'state_name', 'metro'], inplace=True)

#Adding `year` column
il_metro_15_df['il_fiscal_year'] = 2015

#Replacing the 2015 'hud_area_code' for Los Angeles-Long Beach, CA HUD Metro FMR Area since the CBSA code (31080) is consistent in the AHS but changes in the Income Limits data (2015 = METRO31100MM4480 but '17, '19, '21, '23 = METRO31080MM4480. 
#This will ensure correct merging later in the notebook
il_metro_15_df['CBSASub'] = il_metro_15_df['CBSASub'].replace('METRO31100MM4480', 'METRO31080MM4480')

#Moving `year` to 3rd column (index position 2)
il_metro_15_df = il_metro_15_df[[*il_metro_15_df.columns[:2], 'il_fiscal_year', *il_metro_15_df.columns[2:-1]]]

#Rename columns to match API-requested DataFrame
il_metro_15_df.rename(columns={'CBSASub': 'hud_area_code', 'Metro_Area_Name': 'hud_area_name', 'median2015': 'median_income'}, 
                      inplace=True)
il_metro_15_df

hud_area_code                                      hud_area_name  \
0   METRO38060M38060                      Phoenix-Mesa-Glendale, AZ MSA   
1   METRO31080MM4480      Los Angeles-Long Beach, CA HUD Metro FMR Area   
2   METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
3   METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
4   METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
5   METRO33100MM5000   Miami-Miami Beach-Kendall, FL HUD Metro FMR Area   
6   METRO12060M12060  Atlanta-Sandy Springs-Marietta, GA HUD Metro F...   
7   METRO16980M16980   Chicago-Joliet-Naperville, IL HUD Metro FMR Area   
8   METRO14460MM1120  Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area   
9   METRO19820M19820      Detroit-Warren-Livonia, MI HUD Metro FMR Area   
10  METRO35620MM5600                    New York, NY HUD Metro FMR Area   
11  METRO37980M37980    Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA   
12  METRO19100M19100                      Dallas, TX HUD Metro FMR Area   
13  METRO26420M26420  Houston-Baytown-Sugar Land, TX HUD Metro FMR Area   
14  METRO42660MM7600            Seattle-Bellevue, WA HUD Metro FMR Area   

    il_fiscal_year  median_income  l50_1  l50_2  l50_3  l50_4  l50_5  l50_6  \
0             2015          64000  22400  25600  28800  32000  34600  37150   
1             2015          63000  29050  33200  37350  41500  44850  48150   
2             2015          60500  21750  24850  27950  31050  33550  36050   
3             2015         101900  41050  46900  52750  58600  63300  68000   
4             2015         109200  38250  43700  49150  54600  59000  63350   
5             2015          49900  23700  27100  30500  33850  36600  39300   
6             2015          68300  23900  27300  30700  34100  36850  39600   
7             2015          76000  26600  30400  34200  38000  41050  44100   
8             2015          98500  34500  39400  44350  49250  53200  57150   
9             2015          67700  23700  27100  30500  33850  36600  39300   
10            2015          63700  30250  34550  38850  43150  46650  50100   
11            2015          81100  28400  32450  36500  40550  43800  47050   
12            2015          70400  24650  28200  31700  35200  38050  40850   
13            2015          69300  24300  27750  31200  34650  37450  40200   
14            2015          89600  31400  35850  40350  44800  48400  52000   

    ...  ELI_7  ELI_8  l80_1  l80_2  l80_3  l80_4   l80_5   l80_6   l80_7  \
0   ...  36730  40890  35850  41000  46100  51200   55300   59400   63500   
1   ...  36730  40890  46500  53150  59800  66400   71750   77050   82350   
2   ...  36730  40890  34800  39800  44750  49700   53700   57700   61650   
3   ...  43600  46400  65700  75100  84500  93850  101400  108900  116400   
4   ...  40650  43250  47600  54400  61200  68000   73450   78900   84350   
5   ...  36730  40890  37950  43350  48750  54150   58500   62850   67150   
6   ...  36730  40890  38200  43650  49100  54550   58950   63300   67650   
7   ...  36730  40890  42600  48650  54750  60800   65700   70550   75400   
8   ...  36730  40890  48800  55800  62750  69700   75300   80900   86450   
9   ...  36730  40890  37950  43350  48750  54150   58500   62850   67150   
10  ...  36730  40890  48350  55250  62150  69050   74600   80100   85650   
11  ...  36730  40890  45450  51950  58450  64900   70100   75300   80500   
12  ...  36730  40890  39450  45050  50700  56300   60850   65350   69850   
13  ...  36730  40890  38850  44400  49950  55450   59900   64350   68800   
14  ...  36730  40890  46100  52650  59250  65800   71100   76350   81600   

     l80_8  
0    67600  
1    87650  
2    65650  
3   123900  
4    89800  
5    71500  
6    72050  
7    80300  
8    92050  
9    71500  
10   91150  
11   85700  
12   74350  
13   73200  
14   86900  

[15 rows x 28 columns]

#### Stacking the income limit DataFrames

In [29]:
stacked_il_df = pd.concat([il_metro_15_df, metro_staked_il_df], ignore_index=True)

In [30]:
#Exmaining the DataFrame
stacked_il_df

hud_area_code                                      hud_area_name  \
0   METRO38060M38060                      Phoenix-Mesa-Glendale, AZ MSA   
1   METRO31080MM4480      Los Angeles-Long Beach, CA HUD Metro FMR Area   
2   METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
3   METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
4   METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
..               ...                                                ...   
70  METRO42660MM7600            Seattle-Bellevue, WA HUD Metro FMR Area   
71  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
72  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
73  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
74  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   

   il_fiscal_year  median_income  l50_1  l50_2  l50_3  l50_4  l50_5  l50_6  \
0            2015          64000  22400  25600  28800  32000  34600  37150   
1            2015          63000  29050  33200  37350  41500  44850  48150   
2            2015          60500  21750  24850  27950  31050  33550  36050   
3            2015         101900  41050  46900  52750  58600  63300  68000   
4            2015         109200  38250  43700  49150  54600  59000  63350   
..            ...            ...    ...    ...    ...    ...    ...    ...   
70           2023         146500  47950  54800  61650  68500  74000  79500   
71           2017         110300  38650  44150  49650  55150  59600  64000   
72           2019         121300  42500  48550  54600  60650  65550  70400   
73           2021         129000  45150  51600  58050  64500  69700  74850   
74           2023         152100  52750  60300  67850  75350  81400  87450   

    ...  ELI_7  ELI_8  l80_1  l80_2  l80_3   l80_4   l80_5   l80_6   l80_7  \
0   ...  36730  40890  35850  41000  46100   51200   55300   59400   63500   
1   ...  36730  40890  46500  53150  59800   66400   71750   77050   82350   
2   ...  36730  40890  34800  39800  44750   49700   53700   57700   61650   
3   ...  43600  46400  65700  75100  84500   93850  101400  108900  116400   
4   ...  40650  43250  47600  54400  61200   68000   73450   78900   84350   
..  ...    ...    ...    ...    ...    ...     ...     ...     ...     ...   
70  ...  51000  54300  70650  80750  90850  100900  109000  117050  125150   
71  ...  41050  43700  52550  60050  67550   75050   81100   87100   93100   
72  ...  45150  48050  54350  62100  69850   77600   83850   90050   96250   
73  ...  48000  51100  57650  65850  74100   82300   88900   95500  102100   
74  ...  56050  59700  66750  76250  85800   95300  102950  110550  118200   

     l80_8  
0    67600  
1    87650  
2    65650  
3   123900  
4    89800  
..     ...  
70  133200  
71   99100  
72  102450  
73  108650  
74  125800  

[75 rows x 28 columns]

In [31]:
#Inspecting the shape/info of the stacked DataFrame
#Should have 75 rows
print(stacked_il_df.shape)
print(stacked_il_df.info())

(75, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75 entries, 0 to 74
Data columns (total 28 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   hud_area_code   75 non-null     object
 1   hud_area_name   75 non-null     object
 2   il_fiscal_year  75 non-null     object
 3   median_income   75 non-null     int64 
 4   l50_1           75 non-null     int64 
 5   l50_2           75 non-null     int64 
 6   l50_3           75 non-null     int64 
 7   l50_4           75 non-null     int64 
 8   l50_5           75 non-null     int64 
 9   l50_6           75 non-null     int64 
 10  l50_7           75 non-null     int64 
 11  l50_8           75 non-null     int64 
 12  ELI_1           75 non-null     int64 
 13  ELI_2           75 non-null     int64 
 14  ELI_3           75 non-null     int64 
 15  ELI_4           75 non-null     int64 
 16  ELI_5           75 non-null     int64 
 17  ELI_6           75 non-null     int64 
 18  ELI

Need to convert `'il_fiscal_year'` to and `int64` but need to make sure there aren't any null/NaN values in the column

In [32]:
#checking NaN values in 'il_fiscal_uear' column
stacked_il_df['il_fiscal_year'].isna().sum()

0

In [33]:
#coverting 'il_fiscal_year' into an integer
stacked_il_df['il_fiscal_year'] = stacked_il_df['il_fiscal_year'].astype(int)

In [34]:
print(stacked_il_df.shape)
print(stacked_il_df.info())

(75, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75 entries, 0 to 74
Data columns (total 28 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   hud_area_code   75 non-null     object
 1   hud_area_name   75 non-null     object
 2   il_fiscal_year  75 non-null     int64 
 3   median_income   75 non-null     int64 
 4   l50_1           75 non-null     int64 
 5   l50_2           75 non-null     int64 
 6   l50_3           75 non-null     int64 
 7   l50_4           75 non-null     int64 
 8   l50_5           75 non-null     int64 
 9   l50_6           75 non-null     int64 
 10  l50_7           75 non-null     int64 
 11  l50_8           75 non-null     int64 
 12  ELI_1           75 non-null     int64 
 13  ELI_2           75 non-null     int64 
 14  ELI_3           75 non-null     int64 
 15  ELI_4           75 non-null     int64 
 16  ELI_5           75 non-null     int64 
 17  ELI_6           75 non-null     int64 
 18  ELI

#### Creating new column `'cbsa_code'` that extracts the 5-digit cbsa code from the `'hud_area_code'` so I can merge it to the AHS `metro_ami_panel_df` DataFrame.

In [35]:
stacked_il_df['OMB13CBSA'] = stacked_il_df['hud_area_code'].str.extract(r'(\d{5})')

#Looking at the DataFrame to see if the Column was created correctly
print(stacked_il_df)

       hud_area_code                                      hud_area_name  \
0   METRO38060M38060                      Phoenix-Mesa-Glendale, AZ MSA   
1   METRO31080MM4480      Los Angeles-Long Beach, CA HUD Metro FMR Area   
2   METRO40140M40140           Riverside-San Bernardino-Ontario, CA MSA   
3   METRO41860MM7360               San Francisco, CA HUD Metro FMR Area   
4   METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
..               ...                                                ...   
70  METRO42660MM7600            Seattle-Bellevue, WA HUD Metro FMR Area   
71  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
72  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
73  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   
74  METRO47900M47900  Washington-Arlington-Alexandria, DC-VA-MD HUD ...   

    il_fiscal_year  median_income  l50_1  l50_2  l50_3  l50_4  l50_5  l50_6  \
0             2015  

In [36]:
stacked_il_df['OMB13CBSA'].dtype

dtype('O')

#### Merging the AHS and IL datasets so new `'AMI'` variable can be computed and renter households can be catagorized by AMI based on the household income (`'HINCP'`).

In [37]:
#Merging AHS and IL DataFrames
metro_ami_panel_df = pd.merge(
    ahs_panel_df,
    stacked_il_df,
    left_on=['OMB13CBSA', 'SRVYEAR'],
    right_on=['OMB13CBSA', 'il_fiscal_year'],
    how='left',  #use left since I only want records from AHS
    indicator=True
)

### Validating that the join worked correctly
1. Checking `_merge` column to see whether each row came from the left, right, or both DataFrames. All or most values should be `both`.
2. Check for `NaN` values in a colum that should always be present if the merge worked. A result near `0.0` means that all values are present.
3. Cross-tabulate year values to spot mismatches
4. Inspect unmatched rows - look at examples of merge failures

In [38]:
#Checking `_merge` column
metro_ami_panel_df['_merge'].value_counts()

_merge
both          173009
left_only     142470
right_only         0
Name: count, dtype: int64

This equals 315,479, same length as the `ahs_panel_df`.

In [39]:
#Checking `NaN` values
metro_ami_panel_df['median_income'].isna().mean()

0.4515989970806298

In [40]:
#Cross-tab year values to spot mismatches
pd.crosstab(metro_ami_panel_df['SRVYEAR'], metro_ami_panel_df['_merge'])

_merge   left_only   both
SRVYEAR                  
2015         31135  37415
2017         29547  36434
2019         27281  35164
2021         29137  34297
2023         25370  29699

In [41]:
#Inspecting some of the unmatched rows
metro_ami_panel_df[metro_ami_panel_df['_merge'] != 'both'].head(25)

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
0   11000002  1600     -6        -6       8     99998  813.89  03      -6   
1   11000003   840      2        -6       8     99998  581.10  03    2023   
2   11000005    -6      1        -6      -6     99998 7335.97  02    1995   
3   11000006    -6      1        -6      -6     99998 6562.87  02    2019   
4   11000008   800      2        -6       8     99998 1490.80  06    2019   
5   11000009    -6     -6        -6      -6     99998 2210.13  02      -6   
6   11000010  1400      2        -6       1     99998  234.19  02    2020   
10  11000019  1200      2        -6       1     99998  710.49  09    2022   
11  11000021  3000      2        -6       8     99998 5647.69  04    2018   
12  11000025   250      2        -6       1     99998  609.41  06    2018   
13  11000028     0     -6        -6       1     99998  412.45  05      -6   
14  11000030   950     -6        -6       8     99998 4221.44  03      -6   
17  11000035   340      2        -6       1     99998  579.87  07    2016   
18  11000039   950     -6        -6      -9     99998  575.02  03      -6   
19  11000040  1200     -6        -6       8     99998  373.78  04      -6   
20  11000049    -6     -6        -6      -6     99999 2172.87  01      -6   
21  11000053    -6      1        -6      -6     99998 7335.97  02    2013   
23  11000055    -6      1        -6      -6     99998 5515.22  02    2002   
24  11000057    -6     -6        -6      -6     99998 2210.13  02      -6   
25  11000061   870      2        -6       5     99998 7891.15  04    2022   
26  11000066   330     -6        -6       1     99998 2210.13  01      -6   
27  11000068    -6      1        -6      -6     99999 5145.17  02    2002   
28  11000069  2700      2        -6       8     99998 5647.69  04    2019   
29  11000071   950      2        -6       8     99998  709.90  05    2022   
30  11000073    -6      1        -6      -6     99998 7335.97  02    1989   

    NUMPEOPLE  ... ELI_8  l80_1  l80_2  l80_3 l80_4 l80_5  l80_6  l80_7  \
0          -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
1           3  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
2           2  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
3           3  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
4           1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
5          -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
6           1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
10          1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
11          9  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
12          2  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
13         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
14         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
17          1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
18         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
19         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
20         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
21          1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
23          4  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
24         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
25          1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
26         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
27          5  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
28          3  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
29          2  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   
30          1  ...   NaN    NaN    NaN    NaN   NaN   NaN    NaN    NaN   

    l80_8     _merge  
0     NaN  left_only  
1    

In [42]:
#checking NaN values in 'OMB13CBSA' column
# Count NaN values for each selected column within each OMB13CBSA group
nan_counts_by_cbsa = (
    metro_ami_panel_df
    .groupby('OMB13CBSA')[[
        'hud_area_code', 'hud_area_name', 'il_fiscal_year', 'median_income',
        'l50_1', 'l50_2', 'l50_3', 'l50_4', 'l50_5', 'l50_6', 'l50_7', 'l50_8',
        'ELI_1', 'ELI_2', 'ELI_3', 'ELI_4', 'ELI_5', 'ELI_6', 'ELI_7', 'ELI_8',
        'l80_1', 'l80_2', 'l80_3', 'l80_4', 'l80_5', 'l80_6', 'l80_7', 'l80_8'
    ]]
    .apply(lambda df: df.isna().sum())
)

# View the result
with pd.option_context('display.max_columns', None):
    display(nan_counts_by_cbsa)  

hud_area_code  hud_area_name  il_fiscal_year  median_income  \
OMB13CBSA                                                                
12060                  0              0               0              0   
14460                  0              0               0              0   
16980                  0              0               0              0   
19100                  0              0               0              0   
19820                  0              0               0              0   
26420                  0              0               0              0   
31080                  0              0               0              0   
33100                  0              0               0              0   
35620                  0              0               0              0   
37980                  0              0               0              0   
38060                  0              0               0              0   
40140                  0              0               0              0   
41860                  0              0               0              0   
42660                  0              0               0              0   
47900                  0              0               0              0   
99998             110821         110821          110821         110821   
99999              31649          31649           31649          31649   

            l50_1   l50_2   l50_3   l50_4   l50_5   l50_6   l50_7   l50_8  \
OMB13CBSA                                                                   
12060           0       0       0       0       0       0       0       0   
14460           0       0       0       0       0       0       0       0   
16980           0       0       0       0       0       0       0       0   
19100           0       0       0       0       0       0       0       0   
19820           0       0       0       0       0       0       0       0   
26420           0       0       0       0       0       0       0       0   
31080           0       0       0       0       0       0       0       0   
33100           0       0       0       0       0       0       0       0   
35620           0       0       0       0       0       0       0       0   
37980           0       0       0       0       0       0       0       0   
38060           0       0       0       0       0       0       0       0   
40140           0       0       0       0       0       0       0       0   
41860           0       0       0       0       0       0       0       0   
42660           0       0       0       0       0       0       0       0   
47900           0       0       0       0       0       0       0       0   
99998      110821  110821  110821  110821  110821  110821  110821  110821   
99999       31649   31649   31649   31649   31649   31649   31649   31649   

            ELI_1   ELI_2   ELI_3   ELI_4   ELI_5   ELI_6   ELI_7   ELI_8  \
OMB13CBSA                                                                   
12060           0       0       0       0       0       0       0       0   
14460           0       0       0       0       0       0       0       0   
16980           0       0       0       0       0       0       0       0   
19100           0       0       0       0       0       0       0       0   
19820           0       0       0       0       0       0       0       0   
26420           0       0       0       0       0       0       0       0   
31080           0       0       0       0       0       0       0       0   
33100           0       0       0       0       0       0       0       0   
35620           0       0       0       0       0       0       0       0   
37980           0       0       0       0       0       0       0       0   
38060           0       0       0       0       0       0       0       0   
40140           0       0       0       0       0       0       0       0   
41860           0       0       0       0       0       0     

### Checking a random group of owner/renter-occupied housing units in the top 15 largest metropolitan areas to see if merge worked correctly

In [43]:
cbsa_values = [
    '12060', '14460', '16980', '19100', '19820',
    '26420', '31080', '33100', '35620', '37980',
    '38060', '40140', '41860', '42660', '47900'
]

# Filter your dataframe to include only the specified CBSAs
filtered_df = metro_ami_panel_df[metro_ami_panel_df['OMB13CBSA'].isin(cbsa_values)]

# Randomly sample 25 rows from the filtered dataframe
sample_df = filtered_df.sample(n=25, random_state=4)

# Display or work with the result
with pd.option_context('display.max_columns', None):
    display(sample_df)  

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
29428   11053729   800      2         1       2     35620  576.98  09    2014   
106346  11079367    -6     -6        -6      -6     16980 1225.66  02      -6   
268014  11026479  1200     -6        -6       8     33100  848.15  09      -6   
282537  11044598  2000     -6         2       8     31080 1486.76  07      -6   
48762   11088617    -6      1        -6      -6     40140 1071.99  02    2018   
247739  11001037  2700      2         2       8     41860  705.18  07    2012   
106608  11079774  2400      2         2       8     41860  791.67  05    2008   
157566  11056577    -6      1        -6      -6     42660  633.94  02    1977   
35349   11064142  3000     -6         2       8     35620 2920.11  09      -6   
37851   11068792  1100      2        -6       8     40140  958.16  05    2019   
76047   11033162    -6      1        -6      -6     14460 1584.35  04    2011   
19737   11036050    -6      1        -6      -6     42660  708.64  01    2013   
274560  11034746    -6      1        -6      -6     14460  996.98  02    2014   
165755  11067798    -6     -6        -6      -6     35620 2564.48  09      -6   
254302  11009301   450      2        -6       8     26420  987.83  07    2011   
264059  11021514    -6      1        -6      -6     38060  798.93  02    2001   
159009  11058573    -6      1        -6      -6     16980 1274.45  02    2017   
79333   11038164    -6      1        -6      -6     16980 2979.73  02    2013   
92094   11057990    -6      1        -6      -6     35620 3667.50  02    1978   
161205  11061567   740      2        -6       8     19820  849.66  02    2014   
79178   11037909    -6      1        -6      -6     31080 2432.10  02    1970   
64329   11014701  1600     -6        -6       8     19100 1011.22  02      -6   
94822   11062171  1300      2        -6       8     12060 1527.58  08    2021   
78555   11037005    -6      1        -6      -6     19100 1837.29  02    2013   
25897   11047177  1000      2        -6       8     16980 2291.17  06    2017   

        NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY  \
29428           3        3     1950         3         -6      2      -6   
106346         -6       -6     1940         2         25     -6      07   
268014         -6      NaN     2000         2          2     -6      01   
282537         -6      NaN     1930         1         -6     -6      01   
48762           3        1     2018         4         -6     -6      -6   
247739          1      NaN     2000         2         -6      3      -6   
106608          4        2     1940         2         -6      3      -6   
157566          1        4     1970         5         -6     -6      -6   
35349          -6       -6     1960         1         -6     -6      02   
37851           2        1     1930         1         -6      3      -6   
76047           1        4     1919         2         -6     -6      -6   
19737           2        1     1990         3         -6     -6      -6   
274560          2      NaN     1920         4         -6     -6      -6   
165755         -6       -6     1960         3         -6     -6      06   
254302          3      NaN     1980         1         -6      3      -6   
264059          2      NaN     1990         4         -6     -6      -6   
159009          2        1     1950         3         -6     -6      -6   
79333           3        1     1919         2         -6     -6      -6   
92094           2        1     1950         4         -6     -6      -6   
161205          3        4     1940         4         -6      3      -6   
79178           2        1     1960         5         -6     -6      -6   
64329          -6       -6     1960         4         -6     -6      01   
94822           2        3     2000         2         -6      3      -6   
78555           1        1     1960         4         -6     -6      -6   
25897           4        1 

In [44]:
#Checking the shape/info of the new metro_ami_panel_df DataFrame
print(metro_ami_panel_df.shape)
print(metro_ami_panel_df.info())

(315479, 48)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 315479 entries, 0 to 315478
Data columns (total 48 columns):
 #   Column          Non-Null Count   Dtype   
---  ------          --------------   -----   
 0   CONTROL         315479 non-null  object  
 1   RENT            315479 non-null  int64   
 2   TENURE          315479 non-null  object  
 3   RENTCNTRL       315479 non-null  object  
 4   RENTSUB         315479 non-null  object  
 5   OMB13CBSA       315479 non-null  object  
 6   WEIGHT          315479 non-null  float64 
 7   BLD             315479 non-null  object  
 8   HHMOVE          315479 non-null  int64   
 9   NUMPEOPLE       315479 non-null  int64   
 10  SAMEHHLD        246929 non-null  object  
 11  YRBUILT         315479 non-null  int64   
 12  BEDROOMS        315479 non-null  int64   
 13  VACMONTHS       315479 non-null  int64   
 14  HUDSUB          315479 non-null  object  
 15  VACANCY         315479 non-null  object  
 16  HINCP           315479 no

All obversations/rows were matched. Dropping the `'_merge'` column.

In [45]:
metro_ami_panel_df.drop(columns=['_merge'], inplace=True)

In [46]:
#Checking the DataFrame again
with pd.option_context('display.max_columns', None):
    display(metro_ami_panel_df.head())  

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
0  11000002  1600     -6        -6       8     99998  813.89  03      -6   
1  11000003   840      2        -6       8     99998  581.10  03    2023   
2  11000005    -6      1        -6      -6     99998 7335.97  02    1995   
3  11000006    -6      1        -6      -6     99998 6562.87  02    2019   
4  11000008   800      2        -6       8     99998 1490.80  06    2019   

   NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY   HINCP  \
0         -6       -6     1980         4          0     -6      02      -6   
1          3        4     1970         2         -6      3      -6   48000   
2          2        1     1970         4         -6     -6      -6  292500   
3          3        4     1970         3         -6     -6      -6   56000   
4          1        4     1920         1         -6      3      -6   36000   

   TOTHCAMT  SRVYEAR hud_area_code hud_area_name  il_fiscal_year  \
0        -6     2023           NaN           NaN             NaN   
1      1093     2023           NaN           NaN             NaN   
2       810     2023           NaN           NaN             NaN   
3       489     2023           NaN           NaN             NaN   
4       845     2023           NaN           NaN             NaN   

   median_income  l50_1  l50_2  l50_3  l50_4  l50_5  l50_6  l50_7  l50_8  \
0            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4            NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

   ELI_1  ELI_2  ELI_3  ELI_4  ELI_5  ELI_6  ELI_7  ELI_8  l80_1  l80_2  \
0    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
1    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
2    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
3    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
4    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

   l80_3  l80_4  l80_5  l80_6  l80_7  l80_8  
0    NaN    NaN    NaN    NaN    NaN    NaN  
1    NaN    NaN    NaN    NaN    NaN    NaN  
2    NaN    NaN    NaN    NaN    NaN    NaN  
3    NaN    NaN    NaN    NaN    NaN    NaN  
4    NaN    NaN    NaN    NaN    NaN    NaN

In [47]:
#Verify `CONTROL` is unique by survey year
grouped = metro_ami_panel_df.groupby('SRVYEAR')

for year, group in grouped:
    assert group['CONTROL'].nunique() == len(group), f"Control numbers are not unique for SRVYEAR {year}!"

In [48]:
#observing values of the `hud_area_name` column at a list
hud_area_name_list = metro_ami_panel_df['hud_area_name'].unique()

#observing unique CBSA codes
hud_area_code_list = metro_ami_panel_df['hud_area_code'].unique()
OMB13CBSA_list = metro_ami_panel_df['OMB13CBSA'].unique()

# Print the list
print(hud_area_name_list)
print(hud_area_code_list)
print(OMB13CBSA_list)

[nan 'Philadelphia-Camden-Wilmington, PA-NJ-DE-MD MSA'
 'Washington-Arlington-Alexandria, DC-VA-MD HUD Metro FMR Area'
 'New York, NY HUD Metro FMR Area'
 'Boston-Cambridge-Quincy, MA-NH HUD Metro FMR Area'
 'San Francisco, CA HUD Metro FMR Area'
 'Houston-The Woodlands-Sugar Land, TX HUD Metro FMR Area'
 'Miami-Miami Beach-Kendall, FL HUD Metro FMR Area'
 'Atlanta-Sandy Springs-Roswell, GA HUD Metro FMR Area'
 'Phoenix-Mesa-Scottsdale, AZ MSA'
 'Chicago-Joliet-Naperville, IL HUD Metro FMR Area'
 'Dallas, TX HUD Metro FMR Area'
 'Detroit-Warren-Livonia, MI HUD Metro FMR Area'
 'Seattle-Bellevue, WA HUD Metro FMR Area'
 'Los Angeles-Long Beach-Glendale, CA HUD Metro FMR Area'
 'Riverside-San Bernardino-Ontario, CA MSA'
 'Houston-Baytown-Sugar Land, TX HUD Metro FMR Area'
 'Atlanta-Sandy Springs-Marietta, GA HUD Metro FMR Area'
 'Phoenix-Mesa-Glendale, AZ MSA'
 'Los Angeles-Long Beach, CA HUD Metro FMR Area']
[nan 'METRO37980M37980' 'METRO47900M47900' 'METRO35620MM5600'
 'METRO14460MM112

Since some Metro names change within the panel data, I will refer to the CBSA Codes (`'OMB13CBSA'`) for metro-based analyses.

## Categorizing Renter Households based on Area Median Income (AMI)/Income Limits.

I am creating a new `AMI` column consisting of catagorical variables (ELI, VLI, LI, and Above LI) that catagorizes each observation into an AMI group. I will include larger households by incresing the 8-person limit by 8% of the 4-person limit for each additional person beyond 8. 


### 1. Creating a function to assign an AMI catagory to renter households that have over eight (8) peoeple in the unit.

_This is from [HUD USER FY23 Income Limit Documentation](https://www.huduser.gov/portal/datasets/il//il23/IncomeLimitsMethodology-FY23.pdf)_

| **Number of Persons in Family and Percentage Adjustments** | 1    | 2    | 3    | 4    | 5     | 6     | 7     | 8     |
|:-----------------------------------------------------------:|:----:|:----:|:----:|:----:|:-----:|:-----:|:-----:|:-----:|
|                                                             | 70%  | 80%  | 90%  | Base | 108%  | 116%  | 124%  | 132%  |


HUD does not include income limits for families with more than eight persons in the printed lists because of space limitations. For each person over eight-persons, the four-person income limit should be multiplied by an additional 8 percent. (For example, the nine-person limit equals 140 percent \[132 + 8\] of the relevant four-person income limit.) HUD rounds income limits up to the nearest &#36;50.

**Note:**  HUD has used the very low-income limits as the basis for deriving other income limits, unless the relevant statutory language has no references or relationship to low- and very low-income limits, as defined by the U.S. Housing Act of 1937.

#### Example:

* Very Low-Inome limit for 4-person household = &#36;46,750 
* Very Low-Income limit for 8-person household = &#36;61,750 
    * (&#36;46,750 * .32) _rounded to the nearest &#36;50_  
* Low-Income limit for 9-person = &#36;65,450 
    * (&#36;46,750 * .40) _not rounded since the product is a multiple of &#36;50_

In [49]:
#Examinig how many renter/owner-occupied housing units have more than 8 people 
metro_ami_panel_df.groupby('TENURE')['NUMPEOPLE'].value_counts()

TENURE  NUMPEOPLE
-6      -6           41729
1        2           60226
         1           37296
         3           25044
         4           23769
         5           10332
         6            3831
         7            1329
         8             483
         9             200
         10             81
         11             36
         12             22
         13             10
         14              3
         15              2
         16              1
         18              1
2        1           45034
         2           28879
         3           15847
         4           11565
         5            5806
         6            2406
         7             955
         8             362
         9             135
         10             46
         11             24
         12             10
         13              6
         14              3
         15              2
         16              1
         17              1
         18              1
         1

356 owner-occupied housing units have more than 8 people in the unit. 230 renter-occupied housing units have more than 8 people in the unit.

In [50]:
def adjusted_thresholds(threshold_4_array, ppl_count_array):
    """
    Vectorized adjustment of income thresholds for households with more than 8 people.
    - Applies HUD rule: Add 8% of the 4-person threshold for each person beyond 8.
    - Returns NaN for households with 8 or fewer people (which will be handled in the function below)
    
    Parameters:
    - threshold_4_array: array of 4-person income thresholds
    - ppl_count_array: array of household sizes (number of people)
    
    Returns:
    - array of adjusted thresholds (or Nan if 'NUMPEOPLE' <= 8)
    """
    #Convert inputs to NumPy arrays if they are not already an array
    threshold_4_array = np.asarray(threshold_4_array)
    ppl_count_array = np.asarray(ppl_count_array)
    
    #Calculate how many obversations/renter households exceed 8 people
    extra_people = np.maximum(ppl_count_array - 8, 0)
    
    #Compute the total adjustment: 8% increase for each "extra" person and rounds to the nearest $50 
    adjustment = np.round((threshold_4_array * 0.08 * extra_people) / 50) * 50
    
    #Apply the adjusment only if there are extra people; else return NaN
    adjusted_thresholds = np.where(extra_people > 0, threshold_4_array + adjustment, np.nan)
    
    return adjusted_thresholds

### 2. Assigning the correct income threshold based on houhsehold size.

In [51]:
def get_threshold_by_size(df, base_col):
    """
    Return a pandas Series with the appropriate income threshold for each household size ('NUMPEOPLE' 1–8).
    
    Parameters:
    - df: DataFrame containing threshold columns like 'ELI_1' to 'ELI_8', 'l50_1' to l50_8`, etc.
    - base_cal: string, the prefix for the threshold columns ('ELI', 'l50', 'l80', etc.)
    
    Returns:
    - A pandas Series of thresholds matched to each household's NUMPEOPLE.
    """
    #Create list of column names to pull threshold values by size
    columns = [f'{base_col}_{i}' for i in range(1, 9)]
    
    #Convert these threshold columns to a NumPY matrix (2d array)
    matrix = df[columns].to_numpy()
    
    #Calculate the index of the column that corresponds to each row's NUMPEOPLE
    #Clip NUMPEOPLE to 1-8 range, subtract 1 for zero-based indexing
    indexer = df['NUMPEOPLE'].clip(1, 8).astype(int) - 1
    
    #Select the threshold from the appropriate column for each row
    return matrix[np.arange(len(df)), indexer]

### 3. Assign the AMI to each unique renter houhsehold
This is the final part of the function/method for assigning onwer/renter-occupied housing units to an **AMI categories** (Extremely Low Income `ELI`, Very Low Income `VLI`, Low Income `LI`, and Above Low Income `Above LI`) based on:
- AHS Household inomce variable (`'HINCP'`)
- AHS Household size (`'NUMPEOPLE'`)
- HUD Income Limts (`'ELI_1'` to `'ELI_8'`, `'l50_1'` to `'l50_8'`, etc.

In [52]:
def assign_ami_category(df):
    """
    Assign AMI category (ELI, VLI, LI, Above LI) to each unique renter household (`CONTROL`).
    
    Expects Columns:
    - 'HINCP': household income
    - 'NUMPEOPLE': household size
    - Thresholds: 'ELI_1' - 'ELI_8', 'l50_1' - 'l50_8', 'l80_1' - 'l80_8'
    - And 'ELI_4', 'l50_4', 'l80_4' for adjustments for households with more than 8 people
    """
    df = df.copy()
    
    
    #--------- Vectorized thresholds for ELI, VLI, LI ---------
    for prefix, out_col, fallback_col in [
        ('ELI', 'ELI_threshold', 'ELI_4'),
        ('l50', 'VLI_threshold', 'l50_4'),
        ('l80', 'LI_threshold', 'l80_4'),
    ]:
        #Threshold matrix for 1-8 people: shape (n_rows, 8)
        cols = [f'{prefix}_{i}' for i in range (1, 9)]
        matrix = df[cols].to_numpy()
        
        #Index into the threshold matrix using NUMPEOPLE
        ppl_idx = df['NUMPEOPLE'].clip(1, 8).astype(int) - 1
        size_thresholds = matrix[np.arange(len(df)), ppl_idx]
        
        #Adjust for larger households (>8)
        adjusted = adjusted_thresholds(df[fallback_col], df['NUMPEOPLE'])
        
        #Combine adjusted and base thresholds
        df[out_col] = np.where(df['NUMPEOPLE'] <= 8, size_thresholds, adjusted)
        
    #--------- AMI assignment (Fully vectorized) ---------    
    conditions = [
        df['HINCP'] <= df['ELI_threshold'],
        df['HINCP'] <= df['VLI_threshold'],
        df['HINCP'] <= df['LI_threshold']
    ]
    choices = ['ELI', 'VLI', 'LI']    
    df['AMI'] = np.select(conditions, choices, default='Above LI')
    
    # Set AMI = NaN if any of the key inputs is missing
    required_columns = [
        'l50_1', 'l50_2', 'l50_3', 'l50_4', 'l50_5', 'l50_6', 'l50_7', 'l50_8',
        'ELI_1', 'ELI_2', 'ELI_3', 'ELI_4', 'ELI_5', 'ELI_6', 'ELI_7', 'ELI_8',
        'l80_1', 'l80_2', 'l80_3', 'l80_4', 'l80_5', 'l80_6', 'l80_7', 'l80_8',
        'ELI_threshold', 'VLI_threshold', 'LI_threshold'
    ]
    invalid_mask = df[required_columns].isna().any(axis=1) | (df['HINCP'] == -6)
    df.loc[invalid_mask, 'AMI'] = np.nan

    return df

### 4. Using the functions to assign an `'AMI'` category and examining `metro_ami_panel_df` after the funciton was executed.

In [53]:
#Usage
metro_ami_panel_df = assign_ami_category(metro_ami_panel_df)

In [54]:
#Looking at the newly created DataFrame with the 'AMI' values
metro_ami_panel_df.head()

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
0  11000002  1600     -6        -6       8     99998  813.89  03      -6   
1  11000003   840      2        -6       8     99998  581.10  03    2023   
2  11000005    -6      1        -6      -6     99998 7335.97  02    1995   
3  11000006    -6      1        -6      -6     99998 6562.87  02    2019   
4  11000008   800      2        -6       8     99998 1490.80  06    2019   

   NUMPEOPLE  ... l80_3  l80_4  l80_5  l80_6 l80_7 l80_8  ELI_threshold  \
0         -6  ...   NaN    NaN    NaN    NaN   NaN   NaN            NaN   
1          3  ...   NaN    NaN    NaN    NaN   NaN   NaN            NaN   
2          2  ...   NaN    NaN    NaN    NaN   NaN   NaN            NaN   
3          3  ...   NaN    NaN    NaN    NaN   NaN   NaN            NaN   
4          1  ...   NaN    NaN    NaN    NaN   NaN   NaN            NaN   

   VLI_threshold  LI_threshold  AMI  
0            NaN           NaN  NaN  
1            NaN           NaN  NaN  
2            NaN           NaN  NaN  
3            NaN           NaN  NaN  
4            NaN           NaN  NaN  

[5 rows x 51 columns]

In [55]:
#Preliminary review of AMI counts over the survey years.
print(metro_ami_panel_df.groupby(['SRVYEAR', 'TENURE'])['AMI'].value_counts())

SRVYEAR  TENURE  AMI     
2015     1       Above LI    13242
                 LI           2587
                 ELI          2120
                 VLI          1601
         2       Above LI     5026
                 ELI          4258
                 LI           2324
                 VLI          2101
2017     1       Above LI    13064
                 LI           2587
                 ELI          2020
                 VLI          1607
         2       Above LI     5043
                 ELI          3684
                 LI           2377
                 VLI          1963
2019     1       Above LI    12435
                 LI           2618
                 ELI          2079
                 VLI          1608
         2       Above LI     4633
                 ELI          3444
                 LI           2259
                 VLI          1816
2021     1       Above LI    11384
                 LI           2515
                 ELI          2107
                 VLI         

In [56]:
#Looking at number of NaN values in the 'AMI' column
metro_ami_panel_df['AMI'].isna().sum()

162024

In [57]:
#checking NaN values in 'OMB13CBSA' column
# Count NaN values for each selected column within each OMB13CBSA group
nan_ami_counts_by_cbsa = (
    metro_ami_panel_df
    .groupby('OMB13CBSA')[['ELI_threshold', 'VLI_threshold', 'LI_threshold', 'AMI'
    ]]
    .apply(lambda df: df.isna().sum())
)

# View the result
with pd.option_context('display.max_columns', None):
    display(nan_ami_counts_by_cbsa)  

ELI_threshold  VLI_threshold  LI_threshold     AMI
OMB13CBSA                                                    
12060                  0              0             0    1239
14460                  0              0             0     767
16980                  0              0             0    1175
19100                  0              0             0    1152
19820                  0              0             0    1430
26420                  0              0             0    1571
31080                  0              0             0     840
33100                  0              0             0    2757
35620                  0              0             0    1208
37980                  0              0             0    1037
38060                  0              0             0    1648
40140                  0              0             0    1789
41860                  0              0             0     921
42660                  0              0             0     972
47900                  0              0             0    1048
99998             110821         110821        110821  110821
99999              31649          31649         31649   31649

162,024 is the sum of NaNs from the number of observations in `'OMB13CBSA'` column that equal `'99998'` and `'99999'` (142,470) _**plus**_ the sum of NaN values within the 15 largest metropolitan areas that are vacant `'TENURE' == '-6'` (19,554).

### 5. Data Integrity Check

In [58]:
# Data integrity check
def check_data_integrity(df):
    print("🔍 DATA INTEGRITY CHECK\n" + "-"*30)
    
    total_rows = len(df)
    unique_ids = df[['CONTROL', 'SRVYEAR']].drop_duplicates().shape[0]
    control_duplicates = df['CONTROL'].duplicated().sum()
    
    print(f"Total rows: {total_rows}")
    print(f"Unique CONTROL + SRVYEAR pairs: {unique_ids}")
    print(f"Duplicate CONTROL values (across years): {control_duplicates}")
    
    # Nulls in key columns
    missing_income = df['HINCP'].isna().sum()
    missing_size = df['NUMPEOPLE'].isna().sum()
    missing_ami = df['AMI'].isna().sum()
    
    print(f"Missing household income (HINCP): {missing_income}")
    print(f"Missing household size (NUMPEOPLE): {missing_size}")
    print(f"Missing AMI category assignments: {missing_ami}")
    
    # AMI distribution
    print("\n📊 AMI Category Distribution:")
    print(df['AMI'].value_counts(dropna=False).sort_index())
    
    print("\n✅ All checks complete.\n")

In [59]:
#Usage
check_data_integrity(metro_ami_panel_df)

🔍 DATA INTEGRITY CHECK
------------------------------
Total rows: 315479
Unique CONTROL + SRVYEAR pairs: 315479
Duplicate CONTROL values (across years): 223708
Missing household income (HINCP): 0
Missing household size (NUMPEOPLE): 0
Missing AMI category assignments: 162024

📊 AMI Category Distribution:
AMI
Above LI     82514
ELI          30043
LI           23617
VLI          17281
NaN         162024
Name: count, dtype: int64

✅ All checks complete.



In [60]:
#Looking at the instance were `NUMPEOPLE` is greater than 8 people
metro_ami_panel_df.loc[metro_ami_panel_df['NUMPEOPLE'] > 8]

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
11      11000021  3000      2        -6       8     99998 5647.69  04    2018   
686     11001286  1700      2        -6       8     99998  825.56  02    2017   
3642    11006591    50      2         2       6     41860  882.98  02    2007   
4497    11008115    -6      1        -6      -6     99998 4039.95  02    2022   
4722    11008539   350      2        -6       8     99998 5936.33  02    2008   
...          ...   ...    ...       ...     ...       ...     ...  ..     ...   
312024  11081015    -6      1        -6      -6     19820  707.01  02    1996   
313079  11082338    -6      1        -6      -6     99998 3877.67  02    1999   
313211  11082507    -6      1        -6      -6     33100 1138.75  02    2009   
313957  11083428    -6      1        -6      -6     99999 5185.82  02    2006   
314650  11084275  1400      2        -6       8     16980 1653.01  02    2014   

        NUMPEOPLE  ...     l80_3     l80_4     l80_5     l80_6     l80_7  \
11              9  ...       NaN       NaN       NaN       NaN       NaN   
686            11  ...       NaN       NaN       NaN       NaN       NaN   
3642           10  ... 133800.00 148650.00 160550.00 172450.00 184350.00   
4497           10  ...       NaN       NaN       NaN       NaN       NaN   
4722           10  ...       NaN       NaN       NaN       NaN       NaN   
...           ...  ...       ...       ...       ...       ...       ...   
312024         10  ...  48750.00  54150.00  58500.00  62850.00  67150.00   
313079         11  ...       NaN       NaN       NaN       NaN       NaN   
313211         14  ...  48750.00  54150.00  58500.00  62850.00  67150.00   
313957          9  ...       NaN       NaN       NaN       NaN       NaN   
314650         11  ...  54750.00  60800.00  65700.00  70550.00  75400.00   

           l80_8  ELI_threshold  VLI_threshold  LI_threshold       AMI  
11           NaN            NaN            NaN           NaN       NaN  
686          NaN            NaN            NaN           NaN       NaN  
3642   196250.00       64650.00      107750.00     172450.00  Above LI  
4497         NaN            NaN            NaN           NaN       NaN  
4722         NaN            NaN            NaN           NaN       NaN  
...          ...            ...            ...           ...       ...  
312024  71500.00       28150.00       39250.00      62800.00  Above LI  
313079       NaN            NaN            NaN           NaN       NaN  
313211  71500.00       35900.00       50100.00      80150.00       ELI  
313957       NaN            NaN            NaN           NaN       NaN  
314650  80300.00       30050.00       47100.00      75400.00  Above LI  

[586 rows x 51 columns]

### Checking a random group of owner/renter-occupied housing units in the top 15 largest metropolitan areas to see if AMI assignment worked correctly

In [61]:
# Filter your dataframe to include only the specified CBSAs (defined in code line 49)
filtered_ami_df = metro_ami_panel_df[metro_ami_panel_df['OMB13CBSA'].isin(cbsa_values)]

# Randomly sample 25 rows from the filtered dataframe
sample_ami_df = filtered_ami_df.sample(n=25, random_state=12)

# Display or work with the result
with pd.option_context('display.max_columns', None):
    display(sample_ami_df)  

CONTROL  RENT TENURE RENTCNTRL RENTSUB OMB13CBSA  WEIGHT BLD  HHMOVE  \
106217  11079177  1600      2        -6       5     35620 5575.21  08    2010   
138301  11028821    -6      1        -6      -6     35620 5327.57  02    2007   
284394  11046932    -6      1        -6      -6     31080 1648.04  02    1988   
112096  11088402    -6      1        -6      -6     19100 1579.56  02    2017   
85592   11047945  1200      2         1       8     31080 2241.31  05    1987   
153382  11050587  1200      2        -6       5     35620 3571.73  08    1991   
70774   11024948  1300     -6        -6      -9     19100 1011.22  02      -6   
231388  11066575    -6      1        -6      -6     40140  523.87  02    1997   
197990  11022683    -6      1        -6      -6     12060  986.11  02    2015   
306152  11073899  1300      2        -9       8     41860  880.27  08    2003   
94459   11061611    -6      1        -6      -6     19820 1125.87  02    2004   
76988   11034613    -6      1        -6      -6     35620 3413.75  03    2006   
275033  11035316    -6      1        -6      -6     38060  911.60  02    2008   
240641  11078549    -6      1        -6      -6     33100 1127.30  09    2010   
188607  11010248    -6      1        -6      -6     35620 3704.55  03    1991   
262922  11020102    -6      1        -6      -6     38060  711.19  02    1995   
311506  11080379    -6      1        -6      -6     19820  847.66  02    2006   
180552  11088339    -6      1        -6      -6     42660  803.04  02    2019   
296676  11062157    -6      1        -6      -6     12060  960.36  02    2004   
19393   11035445    -6      1        -6      -6     14460 1410.82  02    2015   
79538   11038529    -6      1        -6      -6     26420  778.63  01    2010   
26672   11048541    60      2        -6       1     19100  504.38  09    2019   
122866  11006493  2400      2         2       8     41860  716.26  05    2018   
160581  11060718    -6     -6        -6      -6     19820  307.75  02      -6   
38309   11069588    -6      1        -6      -6     40140  958.16  03    2011   

        NUMPEOPLE SAMEHHLD  YRBUILT  BEDROOMS  VACMONTHS HUDSUB VACANCY  \
106217          2        1     1980         1         -6      3      -6   
138301          3        1     1980         4         -6     -6      -6   
284394          2      NaN     1919         3         -6     -6      -6   
112096          2        3     2016         3         -6     -6      -6   
85592           1        1     1980         2         -6      3      -6   
153382          2        1     1970         2         -6      3      -6   
70774          -6       -6     2000         4         -9     -6      01   
231388          2        3     1970         4         -6     -6      -6   
197990          6        4     1990         3         -6     -6      -6   
306152          4      NaN     1919         2         -6      3      -6   
94459           1        1     1970         3         -6     -6      -6   
76988           3        1     1980         2         -6     -6      -6   
275033          2      NaN     2000         2         -6     -6      -6   
240641          1        2     2000         2         -6     -6      -6   
188607          4        1     1930         3         -6     -6      -6   
262922          1      NaN     1970         3         -6     -6      -6   
311506          4      NaN     1990         3         -6     -6      -6   
180552          4        4     2019         4         -6     -6      -6   
296676          2      NaN     1990         4         -6     -6      -6   
19393           4        1     2000         4         -6     -6      -6   
79538           3        4     1990         4         -6     -6      -6   
26672           3        4     1970         2         -6      1      -6   
122866          2        4     1970         1         -6      3      -6   
160581         -6       -6     1920         4         12     -6      07   
38309           4        1 

Based on reviewing random samples of the data, It appears as though the AMI assignment function worked.

## Exporting `metro_ami_panel_df` to csv for analysis in subsequent Notebooks.

In [ ]:
#Export the DataFrame to a CSV file at the specified location without the index
metro_ami_panel_df.to_csv('data/ahs/metro_ami_panel_df.csv', index=False)